### PR TITLE
Generate events in functions; to avoid non-discrete booleans.

### LINKED ISSUE #4209: Problem with the function setState_phX
I have a library which works well on Modelica 3.3.2, but when converted to Modelica 4.0.0, it doesn't work anymore, and the message error is "Non-real equation in continuous time are not legal:". 

I'm trying to understand the problem by using a very simple example (see below): 
```mo
model model1a
  replaceable package Medium = Modelica.Media.R134a.R134a_ph "Medium model" 
    annotation (choicesAllMatching=true);
  Medium.ThermodynamicState state_a ;
  Modelica.Units.SI.AbsolutePressure p=Medium.p_default;
  Modelica.Units.SI.SpecificEnthalpy h=Medium.h_default;

equation 
  state_a = Medium.setState_phX(Medium.p_default, Medium.h_default);
  //     state_a = Medium.setState_phX(p, h);
end model1a;
```
The above model runs sucessfully. But when I redefine the variable state_a by using the second line in the equation part, the model does not work and the error is : 
```
  Non-real equation in continuous time are not legal:
  state_a.phase = Modelica.Media.R134a.R134a_ph.getPhase_ph(p, h);
```
Any help to understand the error? Many thanks. I use Dymola 2021, Modelica 4.0.0

### PR DESCRIPTION
**Pull request by HansOlsson**
Closes #4209

I don't know if significant enough to push for release.

### CODE REVIEW

File: Modelica/Media/R134a.mo [hunk 1]
```diff
@@ -281,7 +281,7 @@ package R134a "R134a: Medium model for R134a"
       "Set state for pressure and specific enthalpy (X not used since single substance)"
     algorithm
       state := ThermodynamicState(phase=getPhase_ph(p, h), p=p, h=h, d=density_ph(p, h), T=temperature_ph(p, h));
-      annotation (Documentation(info="<html>
+      annotation (Documentation(GenerateEvents=true, info="<html>
```
Comment (beutlich) [inline Modelica/Media/R134a.mo:284]:
```suggestion
      annotation (GenerateEvents=true, Documentation(info="<html>
```

File: Modelica/Media/R134a.mo [hunk 2]
```diff
@@ -2228,7 +2228,7 @@ This function computes the residual helmholtz derivatives of the fundamental equ
       phase := if ((h < hl) or (h > hv) or (p > R134aData.data.FPCRIT)) then 1
          else 2;
 
-      annotation (Documentation(info="<html>
+      annotation (Documentation(GenerateEvents=true, info="<html>
```
Comment (beutlich) [inline Modelica/Media/R134a.mo:2231]:
```suggestion
      annotation (GenerateEvents=true, Documentation(info="<html>
```

**Review by beutlich — CHANGES_REQUESTED** [review]
```mo
model Model1b
  replaceable package Medium = Modelica.Media.R134a.R134a_ph "Medium model"
    annotation (choicesAllMatching=true);
  Medium.ThermodynamicState state_a;
  Modelica.Units.SI.AbsolutePressure p=Medium.p_default;
  Modelica.Units.SI.SpecificEnthalpy h=Medium.h_default;

equation 
  //state_a = Medium.setState_phX(Medium.p_default, Medium.h_default);
  state_a = Medium.setState_phX(p, h);
end Model1b;
```
still fails in Dymola 2024x -  even with the proposed changes. Would be good to have this model as test model.

Are other functions affected as well?

File: Modelica/Media/R134a.mo (hunk 1 above)
Comment (HansOlsson) [reply Modelica/Media/R134a.mo:284]: Oops

**Comment by HansOlsson** [comment]
> ```modelica
> model Model1b
>   replaceable package Medium = Modelica.Media.R134a.R134a_ph "Medium model"
>     annotation (choicesAllMatching=true);
>   Medium.ThermodynamicState state_a;
>   Modelica.Units.SI.AbsolutePressure p=Medium.p_default;
>   Modelica.Units.SI.SpecificEnthalpy h=Medium.h_default;
> 
> equation 
>   //state_a = Medium.setState_phX(Medium.p_default, Medium.h_default);
>   state_a = Medium.setState_phX(p, h);
> end Model1b;
> ```
> 
> still fails in Dymola 2024x - even with the proposed changes. Would be good to have this model as test model.
> 
> Are other functions affected as well?

No, but it seems that I messed up the annotations a bit, it should now be resolved.

**Comment by HansOlsson** [comment]
> still fails in Dymola 2024x - even with the proposed changes. Would be good to have this model as test model.

I added a test-model, but made it similar to an existing one instead.

File: ModelicaTest/Media.mo [hunk 3]
```diff
@@ -950,6 +950,22 @@ is given to compare the approximation.
       annotation (experiment(StopTime=1));
     end R134a_setState_pTX;
 
+    model R134a_setState_phX "Test setState_phX() of R134a"
+      extends Modelica.Icons.Example;
+      replaceable package Medium = Modelica.Media.R134a.R134a_ph "Medium model";
+      SI.Temperature h = Medium.h_default;
+      parameter SI.AbsolutePressure p0 = 10e5 "p at time 0";
+      parameter Real pRate(unit = "Pa/s") = 20e5 "p's rate of change";
```
Comment (beutlich) [inline ModelicaTest/Media.mo:958]:
```suggestion
      parameter SI.PressureSlope pRate = 20e5 "p's rate of change";
```

Comment (HansOlsson) [reply ModelicaTest/Media.mo:958]: It was copied from R134a_setState_pTX which hasn't been updated either. I think we should coordinate that later.

Comment (beutlich) [reply ModelicaTest/Media.mo:958]: Here you go: #4273

**Review by beutlich — APPROVED** [review]
I confirm that Model1b now runs in Dymola 2024x (and still in SimulationX 4.4).

**Comment by beutlich** [comment]
> I don't know if significant enough to push for release.

For me it seems like a MSL bug. What's the opinion of the Media officers?

**Comment by beutlich** [comment]
Unfortunately, the new test model ModelicaTest.Media.TestOnly.R134a_setState_phX raises a [AssignmentFromNonDiscreteToDiscrete](https://doc.simulationx.com/4.4/1033/Content/FAQ/Troubleshooting/ErrorsAndWarnings.htm#Modelica.AssignmentFromNonDiscreteToDiscrete) warning in SimulationX 4.4 and fails to simulate with all tested solvers due to a singular system of equations. It does not matter if this PR is applied or not.

**Comment by HansOlsson** [comment]
> Unfortunately, the new test model ModelicaTest.Media.TestOnly.R134a_setState_phX raises a [AssignmentFromNonDiscreteToDiscrete](https://doc.simulationx.com/4.4/1033/Content/FAQ/Troubleshooting/ErrorsAndWarnings.htm#Modelica.AssignmentFromNonDiscreteToDiscrete) warning in SimulationX 4.4 and fails to simulate with all tested solvers due to a singular system of equations. It does not matter if this PR is applied or not.

Ok, that is problematic. I assume the failure to simulate and assignment-error are related; indicating that the current version of the function doesn't work, and not some unrelated error in the example model. 

I don't know the exact reason - is it related to calling GenerateEvents indirectly (i.e., one function with GenerateEvents calling another one)? If so it might be possible to rewrite it as:
```
  state := ThermodynamicState(phase=
         if ((h < bubbleEnthalpy(SaturationProperties(psat=p,Tsat=0)) 
          or (h > dewEnthalpy(SaturationProperties(psat=p,Tsat=0))) or (p > R134aData.data.FPCRIT)) then 1
         else 2 , p=p, h=h, d=density_ph(p, h), T=temperature_ph(p, h));
```
(I cannot test that in SimulationX.)

So what to do:
- Just add
- Delay until the next release.
- Add the changes in MSL - but don't add the test.
- Use the weird rewriting above - even if horrible copy-pasta. One could argue that this indirect calling isn't fully specified, whereas this rewriting should be ok.

**Comment by hubertus65** [comment]
I would be in favor to add the rewrite proposed by Hans. It makes the event generation explicit. Could you add it to the PR?

**Comment by HansOlsson** [comment]
> I would be in favor to add the rewrite proposed by Hans. It makes the event generation explicit. Could you add it to the PR?

Should be done now. Please re-test

**Review by beutlich — COMMENTED** [review]
@gkurzbach The new test model ModelicaTest.Media.TestOnly.R134a_setState_phX still fails in SimulationX 4.4. Can you please elaborate. Thanks.

**Comment by HansOlsson** [comment]
@gkurzbach can you give some feedback? As I see it, the changes give clear benefits

**Review by casella — APPROVED** [review]
I agree with @hubertus65 that the update by @HansOlsson is good. 

I ran ModelicaTest.Media.TestOnly.R134a_setState_phX with the latest nightly build of OpenModelica and it works fine.

I would definitely add this to 4.1.0

**Comment by casella** [comment]
This PR should be back-ported to maint/4.1.0

**Comment by Esther-Devakirubai** [comment]
Backporting by #4541

**Pull request merged**